### PR TITLE
[To rel/0.13] DoubleWrite for 0.13 version

### PR DESCRIPTION
**Pull request by CRZbulabula**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/service/thrift/impl/TSServiceImpl.java [hunk 1]
```diff
@@ -302,9 +320,92 @@ public TSFetchResultsResp call() throws Exception {
...
+      // create DoubleWriteProtectorService
+      Lock firstLogLock = new ReentrantLock();
+      Condition firstLogCondition = firstLogLock.newCondition();
+      doubleWriteProtectorService =
+          new DoubleWriteProtectorService(firstLogLock, firstLogCondition, doubleWriteSessionPool);
+      new Thread(doubleWriteProtectorService).start();
+      try {
+        firstLogLock.lock();
+        firstLogCondition.await();
+        firstLogLock.unlock();
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/service/thrift/impl/TSServiceImpl.java:366]: unlock should be in finally block

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]
Fantastic work!  Please take a look about some issues~

File: server/src/main/java/org/apache/iotdb/db/doublewrite/DoubleWriteProtectorService.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                doubleWriteSessionPool));
+      }
+    }
+
+    firstLogLock.lock();
+    // create first DoubleWriteLog
+    logWriterLock.lock();
+    createNewLogFile();
+    firstLogCondition.signal();
+    firstLogLock.unlock();
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/doublewrite/DoubleWriteProtectorService.java:105]: This unlock statement is only for thread-safely create the firset DoubleWriteLog.

File: server/src/main/java/org/apache/iotdb/db/service/thrift/impl/TSServiceImpl.java (hunk 1 above)
Comment (CRZbulabula) [reply server/src/main/java/org/apache/iotdb/db/service/thrift/impl/TSServiceImpl.java:366]: This unlock statement is only for thread-safely create the firset DoubleWriteLog.

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 3]
```diff
@@ -926,3 +926,75 @@ timestamp_precision=ms
...
+
+# The transmitting concurrency size of DoubleWrite SessionPool
+# This parameter should better not exceed the sum of concurrent DoubleWriteTask or DoubleWriteConsumer and concurrent DoubleWriteProtector
+# Datatype: int
+# double_write_session_concurrency_size = 8
+
+# The max size per DoubleWriteLog. Default is 1GB
+# Datatype: int
+# double_write_max_log_size = 1073741824
+# The core pool size of DoubleWriteProtector. i.e. The maximum number of running DoubleWriteProtector
```
Comment (qiaojialin) [inline server/src/assembly/resources/conf/iotdb-engine.properties:963]:
```suggestion

# The core pool size of DoubleWriteProtector. i.e. The maximum number of running DoubleWriteProtector
```

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 4]
```diff
@@ -2442,6 +2461,24 @@ private TSDropSchemaTemplateReq getTSDropSchemaTemplateReq(String templateName)
     }
   }
 
+  /** Transmit insert record request for double write */
+  public void doubleWriteTransmit(ByteBuffer buffer)
```
Comment (qiaojialin) [inline session/src/main/java/org/apache/iotdb/session/Session.java:2465]: what's this for?

File: thrift/src/main/thrift/rpc.thrift [hunk 5]
```diff
@@ -416,6 +416,12 @@ struct TSDropSchemaTemplateReq {
   2: required string templateName
 }
 
+struct TSExecuteDoubleWriteReq {
```
Comment (qiaojialin) [inline thrift/src/main/thrift/rpc.thrift:419]:
```suggestion
struct TSInternalSyncWriteReq {
```

File: session/src/main/java/org/apache/iotdb/session/Session.java (hunk 4 above)
Comment (CRZbulabula) [reply session/src/main/java/org/apache/iotdb/session/Session.java:2465]: To transmit all types of the insert plan, and to avoid re-transmit.

**Pull request closed without merging**